src/main/java/th/ac/kmitl/science/comsci/example/nameprinters/NamePrinterLoader.java

```diff
@@ -0,0 +1,56 @@
+package th.ac.kmitl.science.comsci.example.nameprinters;
+
+import org.reflections.Reflections;
+
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public class NamePrinterLoader implements Iterable<NamePrinter> {
+
+    //Singleton implementation
+    private static NamePrinterLoader instance;
+    private final String packageIdentifier;
+    private List<NamePrinter> namePrinters;
+    private NamePrinterLoader() {
+        packageIdentifier = getClass().getPackage().getName();
+
+        try {
+            namePrinters = getAllNamePrinters();
+        } catch (Exception e) {
+            e.printStackTrace();
+            namePrinters = null;
+        }
+    }
+
+    public static NamePrinterLoader getInstance() {
+        if (instance == null)
+            instance = new NamePrinterLoader();
+
+        return instance;
+    }
+
+    @Override
+    public Iterator<NamePrinter> iterator() {
+        return namePrinters.iterator();
+    }
+
+    private List<NamePrinter> getAllNamePrinters() throws Exception {
+        List<NamePrinter> namePrinters = new ArrayList<>();
+        Reflections reflections = new Reflections(packageIdentifier);
+        Set<Class<? extends NamePrinter>> allClasses = reflections.getSubTypesOf(NamePrinter.class);
+
+        for (Class classFound : allClasses) {
+            if (NamePrinter.class.isAssignableFrom(classFound)
+                    && !classFound.isInterface()
+                    && !Modifier.isAbstract(classFound.getModifiers())) {
+                Object namePrinterObj = classFound.newInstance();
+                namePrinters.add((NamePrinter) namePrinterObj);
+            }
+        }
+
+        return namePrinters;
+    }
```

Comment:
The method name is getAllNamePrinters() so it's user should not know any logic about the method error and it should guarantee that it will **always** return list of NamePrinters.

Reply:
@mahasak It's confusing as you mentioned. I renamed it to `getAvailableNamePrinters()` because some `NamePrinter` subclass is abstract and cannot be instantiated.